yarn-project/simulator/src/avm/avm_simulator.test.ts

```diff
@@ -280,6 +280,83 @@ describe('AVM simulator', () => {
+          .mockReturnValue(Promise.resolve(bytecode));
+
+        await new AvmSimulator(context).execute();
+        const results = await new AvmSimulator(context).execute();
+        expect(results.reverted).toBe(false);
+        expect(results.output).toEqual([/*exists=false*/ new Fr(0)]);
+
+        // Nullifier existence check should be in trace
+        expect(context.persistableState.flush().nullifierChecks.length).toEqual(1);
+        expect(context.persistableState.flush().nullifierChecks[0].exists).toEqual(false);
+      });
+      it(`Should execute contract function that checks if a nullifier existence (it does)`, async () => {
+        const utxo = new Fr(42);
+        const calldata = [utxo];
+
+        // Get contract function artifact
+        const artifact = AvmTestContractArtifact.functions.find(f => f.name === 'avm_check_nullifier_exists')!;
+
+        // Decode bytecode into instructions
+        const bytecode = Buffer.from(artifact.bytecode, 'base64');
+
+        const context = initContext({ env: initExecutionEnvironment({ calldata }) });
+        jest
+          .spyOn(context.persistableState.hostStorage.contractsDb, 'getBytecode')
+          .mockReturnValue(Promise.resolve(bytecode));
+
+        // nullifier exists!
+        jest
+          .spyOn(context.persistableState.hostStorage.commitmentsDb, 'getNullifierIndex')
+          .mockReturnValue(Promise.resolve(BigInt(42)));
+
+        await new AvmSimulator(context).execute();
+        const results = await new AvmSimulator(context).execute();
+        expect(results.reverted).toBe(false);
+        expect(results.output).toEqual([/*exists=true*/ new Fr(1)]);
+
+        // Nullifier existence check should be in trace
+        expect(context.persistableState.flush().nullifierChecks.length).toEqual(1);
+        expect(context.persistableState.flush().nullifierChecks[0].exists).toEqual(true);
+      });
+      it(`Should execute contract function that checks emits a nullifier and checks its existence`, async () => {
+        const utxo = new Fr(42);
+        const calldata = [utxo];
+
+        // Get contract function artifact
+        const artifact = AvmTestContractArtifact.functions.find(f => f.name === 'avm_emit_nullifier_and_check')!;
+
+        // Decode bytecode into instructions
+        const bytecode = Buffer.from(artifact.bytecode, 'base64');
+
+        const context = initContext({ env: initExecutionEnvironment({ calldata }) });
+        jest
+          .spyOn(context.persistableState.hostStorage.contractsDb, 'getBytecode')
+          .mockReturnValue(Promise.resolve(bytecode));
+
+        await new AvmSimulator(context).execute();
+        const results = await new AvmSimulator(context).execute();
+        expect(results.reverted).toBe(false);
+
+        // Nullifier existence check should be in trace
+        expect(context.persistableState.flush().newNullifiers).toEqual([utxo]);
```

Comment:
quite hard to read when flush is called each time rather than being called once

Reply:
pulled out